From 06222a7dd9edd835de5dcda12fa386bf44e0f134 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 24 Jan 2010 16:49:23 +0000
Subject: [PATCH] [CONFIGURATION-405] Make sure that the root node of
 XMLPropertyListConfiguration is a PListNode. Otherwise a
 ConfigurationException is thrown when loading configuration files that do not
 have an outer dict element. Ported fix to configuration2 branch.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/branches/configuration2_experimental@902600 13f79535-47bb-0310-9956-ffa450edef68
---
 .../plist/XMLPropertyListConfiguration.java   | 17 ++++++++++++
 .../TestXMLPropertyListConfiguration.java     | 26 +++++++++++++++++++
 src/test/resources/test2.plist.xml            | 13 ++++++++++
 xdocs/changes.xml                             |  4 +++
 4 files changed, 60 insertions(+)
 create mode 100644 src/test/resources/test2.plist.xml

diff --git a/src/main/java/org/apache/commons/configuration2/plist/XMLPropertyListConfiguration.java b/src/main/java/org/apache/commons/configuration2/plist/XMLPropertyListConfiguration.java
index 3a3db216b3..a913305af0 100644
--- a/src/main/java/org/apache/commons/configuration2/plist/XMLPropertyListConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/plist/XMLPropertyListConfiguration.java
@@ -137,6 +137,7 @@ public class XMLPropertyListConfiguration extends AbstractHierarchicalFileConfig
      */
     public XMLPropertyListConfiguration()
     {
+        initRoot();
     }
 
     /**
@@ -227,6 +228,14 @@ public void addProperty(String key, Object value)
 
     public void load(Reader in) throws ConfigurationException
     {
+        // We have to make sure that the root node is actually a PListNode.
+        // If this object was not created using the standard constructor, the
+        // root node is a plain Node.
+        if (!(getRootNode() instanceof PListNode))
+        {
+            initRoot();
+        }
+
         // set up the DTD validation
         EntityResolver resolver = new EntityResolver()
         {
@@ -421,6 +430,14 @@ else if (value != null)
         }
     }
 
+    /**
+     * Helper method for initializing the configuration's root node.
+     */
+    private void initRoot()
+    {
+        setRootNode(new PListNode());
+    }
+
     /**
      * SAX Handler to build the configuration nodes while the document is being parsed.
      */
diff --git a/src/test/java/org/apache/commons/configuration2/plist/TestXMLPropertyListConfiguration.java b/src/test/java/org/apache/commons/configuration2/plist/TestXMLPropertyListConfiguration.java
index f580c89113..757ac39f77 100644
--- a/src/test/java/org/apache/commons/configuration2/plist/TestXMLPropertyListConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/plist/TestXMLPropertyListConfiguration.java
@@ -31,6 +31,7 @@
 import org.apache.commons.configuration2.Configuration;
 import org.apache.commons.configuration2.ConfigurationAssert;
 import org.apache.commons.configuration2.ConfigurationComparator;
+import org.apache.commons.configuration2.ConfigurationException;
 import org.apache.commons.configuration2.FileConfiguration;
 import org.apache.commons.configuration2.InMemoryConfiguration;
 import org.apache.commons.configuration2.StrictConfigurationComparator;
@@ -341,4 +342,29 @@ public void testInitCopy()
         StrictConfigurationComparator comp = new StrictConfigurationComparator();
         assertTrue("Configurations are not equal", comp.compare(config, copy));
     }
+
+    /**
+     * Tests whether a configuration can be loaded that does not start with a
+     * <code>dict</code> element. This test case is related to
+     * CONFIGURATION-405.
+     */
+    public void testLoadNoDict() throws ConfigurationException
+    {
+        XMLPropertyListConfiguration plist = new XMLPropertyListConfiguration();
+        plist.setFile(ConfigurationAssert.getTestFile("test2.plist.xml"));
+        plist.load();
+        assertFalse("Configuration is empty", plist.isEmpty());
+    }
+
+    /**
+     * Tests whether a configuration that does not start with a
+     * <code>dict</code> element can be loaded from a constructor. This test
+     * case is related to CONFIGURATION-405.
+     */
+    public void testLoadNoDictConstr() throws ConfigurationException
+    {
+        XMLPropertyListConfiguration plist = new XMLPropertyListConfiguration(
+                ConfigurationAssert.getTestFile("test2.plist.xml"));
+        assertFalse("Configuration is empty", plist.isEmpty());
+    }
 }
diff --git a/src/test/resources/test2.plist.xml b/src/test/resources/test2.plist.xml
new file mode 100644
index 0000000000..03a25574a6
--- /dev/null
+++ b/src/test/resources/test2.plist.xml
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist SYSTEM "file://localhost/System/Library/DTDs/PropertyList.dtd">
+<plist version="1.0">
+  <array>
+    <array>
+      <string>http://www.google.com/search?hl=en&amp;client=safari&amp;rls=en&amp;q=RFC822MessageDatasPboardType&amp;btnG=Search&amp;aq=f&amp;oq=&amp;aqi=</string>
+    </array>
+    <array>
+      <string>RFC822MessageDatasPboardType - Google Search</string>
+    </array>
+  </array>
+</plist>
+
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 574aa4a8e9..88c4e228cd 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -79,6 +79,10 @@
     </release>
 
     <release version="1.7" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-405">
+        XMLPropertyListConfiguration no longer throws a ConfigurationException
+        if the file to be loaded does not have an outer dict element.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-403">
         When an empty XMLConfiguration was saved and reloaded the root element
         was assigned an empty text value. Because of this isEmpty() returned
